File: services/product-service/download_real_images.py
#!/usr/bin/env python3
"""
Download real product images from Google/Bing and save to database
"""

import requests
import sqlite3
import os
import time
from pathlib import Path
from urllib.parse import urlparse
import json
from datetime import datetime

# Config
API_BASE = "http://localhost:3002"
DB_PATH = "./db/products.db"
IMG_DIR = "./public/images/products"
TIMEOUT = 10

# Create images directory if not exist
Path(IMG_DIR).mkdir(parents=True, exist_ok=True)

def slugify(name):
    """Convert product name to slug"""
    return str(name).lower().replace(' ', '-').replace("'", '').replace('"', '').replace('/', '-')

def search_product_image(product_name, category):
    """
    Search for product image using Bing Image API
    Returns image URL
    """
    search_query = f"{product_name} product official"
    
    try:
        # Using Bing Image Search (more reliable than Google)
        # Note: In production, use proper Bing Search API
        # For now, use a backup approach with DuckDuckGo or direct manufacturer URLs
        
        print(f"  🔍 Searching: {product_name}")
        
        # Try direct manufacturer URLs first
        manufacturer_map = {
            'apple': 'https://www.apple.com/shop/product/',
            'samsung': 'https://www.samsung.com/us/smartphones/',
            'google': 'https://store.google.com/us/product/',
            'dell': 'https://www.dell.com/en-us/shop/laptops/',
            'hp': 'https://www.hp.com/us-en/shop/slp/laptops/',
            'lenovo': 'https://www.lenovo.com/us/en/laptops/',
            'sony': 'https://www.sony.com/electronics/',
            'lg': 'https://www.lg.com/us/tvs-oled-tvs/',
            'asus': 'https://www.asus.com/us/',
            'acer': 'https://www.acer.com/ac/en/US/content/home',
        }
        
        # Extract brand from product name
        product_lower = product_name.lower()
        brand = None
        for key in manufacturer_map:
            if key in product_lower:
                brand = key
                break
        
        # Use placeholder image for now - we'll replace with real download
        return None
        
    except Exception as e:
        print(f"  ❌ Error searching: {e}")
        return None

def download_image(url, filename, timeout=10):
    """Download image from URL and save to file"""
    try:
        print(f"  ⬇️  Downloading to {filename}...")
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
        
        response = requests.get(url, headers=headers, timeout=timeout, verify=False)
        response.raise_for_status()
        
        # Ensure filename has jpg extension
        if not filename.lower().endswith('.jpg'):
            filename = filename.rsplit('.', 1)[0] + '.jpg'
        
        with open(filename, 'wb') as f:
            f.write(response.content)
        
        print(f"  ✅ Saved: {filename}")
        return True
        
    except Exception as e:
        print(f"  ❌ Failed to download: {e}")
        return False

def get_products():
    """Fetch all products from API"""
    try:
        response = requests.get(f"{API_BASE}/", timeout=TIMEOUT)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"❌ Failed to fetch products: {e}")
        return []

def update_product_image(product_id, image_url):
    """Update product image_url in database via API"""
    try:
        payload = {"image_url": image_url}
        response = requests.post(
            f"{API_BASE}/update-image/{product_id}",
            json=payload,
            timeout=TIMEOUT
        )
        response.raise_for_status()
        return True
    except Exception as e:
        print(f"  ❌ Failed to update DB: {e}")
        return False

def get_existing_images():
    """Get list of existing image files"""
    try:
        files = os.listdir(IMG_DIR)
        return [f for f in files if f.lower().endswith('.jpg')]
    except:
        return []

def create_fallback_image(filename, product_name, category):
    """
    Create a fallback SVG image for products without real images
    """
    try:
        # Create a simple SVG placeholder
        colors = {
            'phone': '#1D3F6E',
            'laptop': '#2D4D2D', 
            'accessory': '#5A3A17'
        }
        bg_color = colors.get(category, '#334155')
        
        svg_content = f'''<?xml version="1.0" encoding="UTF-8"?>
<svg width="400" height="400" xmlns="http://www.w3.org/2000/svg">
  <rect width="400" height="400" fill="{bg_color}"/>
  <text x="200" y="200" text-anchor="middle" fill="white" font-size="20" font-family="Arial">
    {product_name[:30]}
  </text>
</svg>'''
        
        # Convert SVG to simple JPEG using PIL if available
        # For now, skip - we'll rely on real downloads
        return None
        
    except Exception as e:
        print(f"  ❌ Failed to create fallback: {e}")
        return None

def main():
    print("\n" + "="*80)
    print("🖼️  REAL PRODUCT IMAGE DOWNLOADER")
    print("="*80)
    
    # Step 1: Get existing images
    existing = get_existing_images()
    print(f"\n📊 Existing images: {len(existing)}")
    
    # Step 2: Get products
    print(f"\n🔄 Fetching products from {API_BASE}...")
    products = get_products()
    
    if not products:
        print("❌ No products found!")
        return
    
    print(f"✅ Found {len(products)} products")
    
    # Step 3: For each product that needs an image
    updated_count = 0
    skipped_count = 0
    
    for idx, product in enumerate(products, 1):
        product_id = product.get('id')
        name = product.get('name', 'Unknown')
        category = product.get('category', 'accessory')
        current_image = product.get('image_url', '')
        
        # Extract slug from current image_url
        slug = None
        if 'product-photo/' in current_image:
            slug = current_image.split('product-photo/')[1].split('?')[0].split('.')[0]
        
        if not slug:
            slug = slugify(name)
        
        img_filename = f"{IMG_DIR}/{slug}.jpg"
        
        # Check if image already exists
        if os.path.exists(img_filename):
            print(f"\n✅ [{idx}/{len(products)}] {name}")
            print(f"   📄 Image exists: {slug}.jpg")
            skipped_count += 1
            continue
        
        print(f"\n⏳ [{idx}/{len(products)}] {name}")
        print(f"   📁 Category: {category}")
        print(f"   🔗 Current: {current_image[-50:] if current_image else 'None'}")
        
        # Try to search and download
        # For now, we'll create a note of what needs to be downloaded
        print(f"   ⚠️  Image not found locally - needs manual mapping")
        
        time.sleep(0.1)  # Small delay between items
    
    print("\n" + "="*80)
    print(f"📈 Summary:")
    print(f"   ✅ Already have: {skipped_count}")
    print(f"   ⏳ Need to download: {len(products) - skipped_count}")
    print("="*80)
    print("\n💡 Next steps:")
    print("   1. Use browser to visit: https://google.com/search?tbm=isch&q=<product+name>")
    print("   2. Download images manually for missing products")
    print("   3. Save as: products/<slug>.jpg")
    print("   4. Run this script again to verify")

if __name__ == "__main__":
    main()

File: services/product-service/download_real_images.js
// Download real product images from Unsplash/Pexels for each product
// Then update database with new image URLs

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const sqlite3 = require('sqlite3');

const IMG_DIR = './public/images/products';
const DB_PATH = './db/products.db';
const API_PORT = 3002;

// Product → Image URL mapping
// Using Unsplash free images (no auth required for moderate use)
const PRODUCT_IMAGES = {
  'iphone-15-pro': 'https://images.unsplash.com/photo-1592286927505-1def25115558?w=600&h=600&fit=crop',
  'nothing-phone-2a': 'https://images.unsplash.com/photo-1506794778202-cad84cf45f1d?w=600&h=600&fit=crop',
  'google-pixel-8-pro': 'https://images.unsplash.com/photo-1598327105666-5b89351aff97?w=600&h=600&fit=crop',
  'oneplus-12': 'https://images.unsplash.com/photo-1511707267537-b85faf00021e?w=600&h=600&fit=crop',
  'macbook-pro-m3': 'https://images.unsplash.com/photo-1517336714731-489689fd1ca8?w=600&h=600&fit=crop',
  'samsung-a54': 'https://images.unsplash.com/photo-1511707267537-b85faf00021e?w=600&h=600&fit=crop',
  'xiaomi-13t': 'https://images.unsplash.com/photo-1506794778202-cad84cf45f1d?w=600&h=600&fit=crop',
  'dell-xps-15': 'https://images.unsplash.com/photo-1527864550417-7fd231fc5205?w=600&h=600&fit=crop',
  'dell-inspiron-gaming': 'https://images.unsplash.com/photo-1527864550417-7fd231fc5205?w=600&h=600&fit=crop',
  'lenovo-ideapad': 'https://images.unsplash.com/photo-1517336714731-489689fd1ca8?w=600&h=600&fit=crop',
  'hp-pavilion': 'https://images.unsplash.com/photo-1517336714731-489689fd1ca8?w=600&h=600&fit=crop',
  'asus-vivobook': 'https://images.unsplash.com/photo-1517336714731-489689fd1ca8?w=600&h=600&fit=crop',
  'acer-swift-go-14': 'https://images.unsplash.com/photo-1527864550417-7fd231fc5205?w=600&h=600&fit=crop',
  'airpods-pro': 'https://images.unsplash.com/photo-1505740420928-5e560c06d30e?w=600&h=600&fit=crop',
  'sony-wh-1000xm5': 'https://images.unsplash.com/photo-1484704849700-f032a568e944?w=600&h=600&fit=crop',
  'logitech-mx-keys': 'https://images.unsplash.com/photo-1587829191301-4c3943b65a58?w=600&h=600&fit=crop',
  'logitech-mx-master-3s': 'https://images.unsplash.com/photo-1527814050087-3793815479db?w=600&h=600&fit=crop',
  'samsung-t7-ssd-1tb': 'https://images.unsplash.com/photo-1597872200969-2b65d56bd16b?w=600&h=600&fit=crop',
  'anker-65w-gan-charger': 'https://images.unsplash.com/photo-1591637281519-b7b05a50abe2?w=600&h=600&fit=crop',
  'razer-blackshark-v2-x': 'https://images.unsplash.com/photo-1505740420928-5e560c06d30e?w=600&h=600&fit=crop',
  'jbl-tune-510bt': 'https://images.unsplash.com/photo-1484704849700-f032a568e944?w=600&h=600&fit=crop',
};

// Additional generic fallbacks by category
const CATEGORY_IMAGES = {
  'phone': 'https://images.unsplash.com/photo-1511707267537-b85faf00021e?w=600&h=600&fit=crop',
  'laptop': 'https://images.unsplash.com/photo-1527864550417-7fd231fc5205?w=600&h=600&fit=crop',
  'accessory': 'https://images.unsplash.com/photo-1587829191301-4c3943b65a58?w=600&h=600&fit=crop',
};

function downloadImage(url, filepath) {
  return new Promise((resolve) => {
    const protocol = url.startsWith('https') ? https : http;
    
    const request = protocol.get(url, { timeout: 5000 }, (response) => {
      // Handle redirects
      if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        downloadImage(response.headers.location, filepath).then(resolve);
        return;
      }
      
      if (response.statusCode !== 200) {
        console.log(`    ❌ HTTP ${response.statusCode}`);
        resolve(false);
        return;
      }
      
      const fileStream = fs.createWriteStream(filepath);
      response.pipe(fileStream);
      
      fileStream.on('finish', () => {
        fileStream.close();
        const size = fs.statSync(filepath).size / 1024;
        console.log(`    ✅ Downloaded (${size.toFixed(1)}KB)`);
        resolve(true);
      });
      
      fileStream.on('error', (err) => {
        fs.unlink(filepath, () => {});
        console.log(`    ❌ Write error: ${err.message}`);
        resolve(false);
      });
    });
    
    request.on('error', (err) => {
      console.log(`    ❌ Download error: ${err.message}`);
      resolve(false);
    });
    
    request.on('timeout', () => {
      request.destroy();
      console.log(`    ❌ Timeout`);
      resolve(false);
    });
  });
}

async function getAllProducts() {
  return new Promise((resolve) => {
    const db = new sqlite3.Database(DB_PATH);
    db.all(
      'SELECT id, name, category FROM products ORDER BY id',
      (err, rows) => {
        db.close();
        resolve(err ? [] : rows);
      }
    );
  });
}

function updateProductImage(productId, imageUrl) {
  return new Promise((resolve) => {
    const db = new sqlite3.Database(DB_PATH);
    db.run(
      'UPDATE products SET image_url = ? WHERE id = ?',
      [imageUrl, productId],
      (err) => {
        db.close();
        resolve(!err);
      }
    );
  });
}

function getImageUrl(slug, category) {
  return `http://localhost:${API_PORT}/images/product-photo/${slug}.jpg?category=${category}&v=20260405c`;
}

async function main() {
  console.log('\n' + '='.repeat(90));
  console.log('🖼️  DOWNLOAD REAL PRODUCT IMAGES');
  console.log('='.repeat(90));

  // Create img directory
  if (!fs.existsSync(IMG_DIR)) {
    fs.mkdirSync(IMG_DIR, { recursive: true });
  }

  // Get products
  console.log('\n📂 Loading products from database...');
  const products = await getAllProducts();

  if (!products.length) {
    console.log('❌ No products found!');
    return;
  }

  console.log(`✅ Found ${products.length} products\n`);

  let downloaded = 0;
  let skipped = 0;
  let failed = 0;

  for (let i = 0; i < products.length; i++) {
    const prod = products[i];
    const slug = prod.name.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '');
    const filepath = path.join(IMG_DIR, `${slug}.jpg`);
    
    // Check if already exists
    if (fs.existsSync(filepath)) {
      const size = fs.statSync(filepath).size / 1024;
      console.log(`✅ [${i + 1}/${products.length}] ${prod.name.padEnd(40)} | ${size.toFixed(1)}KB (exists)`);
      skipped++;
      continue;
    }

    console.log(`⏳ [${i + 1}/${products.length}] ${prod.name.padEnd(40)} | Downloading...`);

    // Get image URL
    let imageUrl = PRODUCT_IMAGES[slug];
    if (!imageUrl) {
      imageUrl = CATEGORY_IMAGES[prod.category] || CATEGORY_IMAGES['accessory'];
    }

    // Download
    const success = await downloadImage(imageUrl, filepath);

    if (success) {
      // Update database
      const dbImageUrl = getImageUrl(slug, prod.category);
      const dbOk = await updateProductImage(prod.id, dbImageUrl);
      
      if (dbOk) {
        console.log(`    📝 Database updated`);
      }
      
      downloaded++;
    } else {
      failed++;
    }
  }

  console.log('\n' + '='.repeat(90));
  console.log('📊 SUMMARY:');
  console.log(`   ✅ Already existing: ${skipped}`);
  console.log(`   ⬇️  Downloaded: ${downloaded}`);
  console.log(`   ❌ Failed: ${failed}`);
  console.log(`   Total: ${downloaded + skipped}/${products.length}`);
  console.log('='.repeat(90));
  
  if (downloaded + skipped === products.length) {
    console.log('✅ All products have images!');
  } else {
    console.log(`⚠️  ${failed} products still need images`);
  }
  console.log('');
}

main().catch(console.error);
